coreos-ostree-importer/coreos_ostree_importer.py

```diff
@@ -0,0 +1,280 @@
+        # lose the traceback logs from the container
+        try:
+            self.process(message)
+            send_message(msg=message.body, status="SUCCESS")
+        except Exception as e:
+            logger.error("Caught Exception!")
+            logger.error("###################################")
+            traceback.print_exc()
+            logger.error("###################################")
+            logger.error("Replying with a FAILURE message...")
+            send_message(msg=message.body, status="FAILURE")
+            logger.error("\t continuing...")
+            pass
+
+    def process(self, message: fedora_messaging.api.Message):
+        logger.debug(message.topic)
+        logger.debug(message.body)
+
+        # Grab the raw message body and parse out pieces
+        msg = message.body
+        basearch = msg["basearch"]
+        build_id = msg["build_id"]
+        checksum = msg["checksum"]
+        commit_url = msg["commit"]
+        ostree_checksum = msg["ostree_checksum"]
+        ostree_ref = msg["ostree_ref"]
+        stream = msg["stream"]
+        target_repo = msg["target_repo"]
+
+        # Qualify arguments
+        if not checksum.startswith("sha256:"):
+            raise Exception("checksum value must start with sha256:")
+        if target_repo not in KNOWN_OSTREE_REPOS.keys():
+            raise Exception(f"Provided target repo is unknown: {target_repo}")
+
+        sha256sum = checksum[7:]
+        target_repo_path = KNOWN_OSTREE_REPOS[target_repo]
+        source_repo_path = None
+
+        # Parse the message
+
+        # Check the target repo to make sure it exists
+        # and is accessible
+        if not ostree_repo_exists(target_repo_path):
+            raise Exception(f"OSTree repo does not exist at {target_repo_path}")
+
+        # Detect if the commit already exists in the target repo
+        if commit_exists(target_repo_path, ostree_checksum):
+            return
+
+        # Import the OSTree commit to the specified repo. We'll use
+        # a temporary directory to untar the repo into.
+        with tempfile.TemporaryDirectory() as tmpdir:
+            # If the target repo is the prod repo the commit could
+            # already have been imported into the compose repo. If it
+            # is already in the compose repo then let's just pull-local
+            # from there to save downloading all from the net again.
+            if target_repo == "prod" and commit_exists(
+                repo=KNOWN_OSTREE_REPOS["compose"], commit=ostree_checksum
+            ):
+                source_repo_path = KNOWN_OSTREE_REPOS["compose"]
+            else:
+                # Grab the file from s3 and then pull local
+                untar_file_from_s3(url=commit_url, tmpdir=tmpdir, sha256sum=sha256sum)
+                source_repo_path = tmpdir
+
+            # one more sanity check: make sure buildid == version
+            assert_commit_has_version(
```

Comment:
Another sanity check we should do is that the latest commit on the target ref is the parent of the new commit we're importing to make sure we're not introducing any holes.

Reply:
+1